src/core/thread/link_metrics.cpp

```diff
@@ -0,0 +1,344 @@
+        pos += sizeof(LinkMetricsTypeIdFlags);
+
+        switch (typeIdFlags.GetMetricEnum())
+        {
+        case OT_LINK_METRICS_PDU_COUNT:
+            VerifyOrExit(typeIdFlags.IsLengthFlagSet(), OT_NOOP);
+            metricsValues.mMetrics.mPduCount = true;
+            aMessage.Read(pos, sizeof(uint32_t), &metricsValues.mPduCountValue);
+            pos += sizeof(uint32_t);
+            otLogDebgMle(" - PDU Counter: %d (Count/Summation)\r\n", metricsValues.mPduCountValue);
+            break;
+
+        case OT_LINK_METRICS_LQI:
+            VerifyOrExit(!typeIdFlags.IsLengthFlagSet(), OT_NOOP);
+            metricsValues.mMetrics.mLqi = true;
+            aMessage.Read(pos, sizeof(uint8_t), &metricsValues.mLqiValue);
+            pos += sizeof(uint8_t);
+            otLogDebgMle(" - LQI: %d (Exponential Moving Average)\r\n", metricsValues.mLqiValue);
+            break;
+
+        case OT_LINK_METRICS_MARGIN:
+            VerifyOrExit(!typeIdFlags.IsLengthFlagSet(), OT_NOOP);
+            metricsValues.mMetrics.mLinkMargin = true;
+            aMessage.Read(pos, sizeof(uint8_t), &metricsValues.mLinkMarginValue);
+            pos += sizeof(uint8_t);
+            otLogDebgMle(" - Margin: %d (Exponential Moving Average)\r\n", metricsValues.mLinkMarginValue);
+            break;
+
+        case OT_LINK_METRICS_RSSI:
+            VerifyOrExit(!typeIdFlags.IsLengthFlagSet(), OT_NOOP);
+            metricsValues.mMetrics.mRssi = true;
+            aMessage.Read(pos, sizeof(uint8_t), &metricsValues.mRssiValue);
+            pos += sizeof(uint8_t);
+            otLogDebgMle(" - RSSI: %d (Exponential Moving Average)\r\n", metricsValue.mRssiValue);
+            break;
+
+        default:
+            break;
+        }
+    }
+
+    mLinkMetricsReportCallback(&aAddress, &metricsValues, mLinkMetricsReportCallbackContext);
+
+exit:
+    return;
+}
+
+void LinkMetrics::SetLinkMetricsReportCallback(otLinkMetricsReportCallback aCallback, void *aCallbackContext)
+{
+    mLinkMetricsReportCallback        = aCallback;
+    mLinkMetricsReportCallbackContext = aCallbackContext;
+}
+
+otError LinkMetrics::SendLinkMetricsQuery(const Ip6::Address &          aDestination,
+                                          uint8_t                       aSeriesId,
+                                          const LinkMetricsTypeIdFlags *aTypeIdFlags,
+                                          uint8_t                       aTypeIdFlagsCount)
+{
+    otError                 error = OT_ERROR_NONE;
+    LinkMetricsQueryId      linkMetricsQueryId;
+    LinkMetricsQueryOptions linkMetricsQueryOptions;
+    uint8_t                 length = 0;
+    static const uint8_t    tlvs[] = {Mle::Tlv::kLinkMetricsReport};
+    uint8_t                 buf[sizeof(Tlv) * 3 + sizeof(LinkMetricsQueryId) + sizeof(LinkMetricsQueryOptions)];
+    Tlv *                   tlv = reinterpret_cast<Tlv *>(buf);
```

Comment:
Isn't possible to use `LinkMetricsQueryTlv` here?

Reply:
I think we can use it but it's kind of misguiding. `LinkMetricsQueryTlv` is for parsing, it contains TLV header (Type and value) and a field for its content. In this function we are trying to write a `LinkMetricsQueryTlv`into a buf. And at line 220 the purpose is just to write a TLV header first. So if we use `LinkMetricsQueryTlv`, it seems like we're setting the buffer by setting `LinkMetricsQueryTlv`'s fields. However we actually set its sub-TLVs separately.